### PR TITLE
WIP: math/bits: use LLVM intrinsics

### PR DESCRIPTION
**Pull request by niaow**
This has some issues with WASM right now, so it still needs work.
Mostly putting this here so that I don't forget about it again.

### CODE REVIEW

**Comment by niaow** [comment]
This is currently blocked by building `compiler-rt` for wasm.

File: compiler/compiler.go [hunk 1]
```diff
@@ -716,6 +716,10 @@ func (c *compilerContext) createPackage(irbuilder llvm.Builder, pkg *ssa.Package
 			// Create the function definition.
 			b := newBuilder(c, irbuilder, member)
 			if member.Blocks == nil {
+				if cd, ok := compDefs[b.llvmFn.Name()]; ok {
+					// This is a compiler-defined function.
+					b.fillCompDef(cd)
+				}
```
Comment (aykevl) [inline compiler/compiler.go:722]: We have lots of compiler-overridden functions already. We currently replace `*ssa.Call` instructions to them with their value inline, but perhaps it makes sense to replace the contents of the functions instead (at least, in some cases). See the `syscall.Syscall*` functions. In any case, I think we should be consistent and do one or the other, I would be fine with changing most of those call replacements to replacing the definition of functions as a refactor.

**Review by aykevl — COMMENTED** [review]
Is there a reason why you copied math/bits? If not absolutely needed, I'd prefer to keep using the upstream version.

**Comment by niaow** [comment]
> Is there a reason why you copied math/bits?

I had done it this way since it required replacing almost every function in `math/bits`.
This can be done in the compiler, but it would effectively become a full rewrite of `math/bits`. . .  but within the compiler.

File: compiler/compiler.go (hunk 1 above)
Comment (niaow) [reply compiler/compiler.go:722]: I think I can turn this into a replacement in this case.

**Comment by niaow** [comment]
Additionally there are some other smaller details like using the runtime's divide-by-zero panic function (the upstream version tries to linkname it, but it works differently in the upstream version).

**Comment by niaow** [comment]
One thing which I just realized is that the rotate functions do not currently have the specified constant-time property on AVR. . .

**Comment by aykevl** [comment]
> > Is there a reason why you copied math/bits?
> 
> I had done it this way since it required replacing almost every function in `math/bits`.

Right, I see. That seems reasonable.

> Additionally there are some other smaller details like using the runtime's divide-by-zero panic function (the upstream version tries to linkname it, but it works differently in the upstream version).

We might want to switch to how upstream does it some day, so that runtime panics can be caught. But that's for a separate PR.

**Comment by dgryski** [comment]
Dup https://github.com/tinygo-org/tinygo/pull/3620

**Pull request closed without merging**